rclcpp/test/test_intra_process_manager.cpp

```diff
@@ -0,0 +1,748 @@
+// limitations under the License.
+
+#include <gtest/gtest.h>
+
+#include <rclcpp/macros.hpp>
+
+// Mock up publisher and subscription base to avoid needing an rmw impl.
+namespace rclcpp { namespace publisher { namespace mock {
+
+class Publisher
+{
+public:
+  RCLCPP_SMART_PTR_DEFINITIONS(Publisher);
+
+  Publisher() : mock_topic_name(""), mock_queue_size(0) {}
+
+  std::string get_topic_name() const {return mock_topic_name;}
+  size_t get_queue_size() const {return mock_queue_size;}
+
+  std::string mock_topic_name;
+  size_t mock_queue_size;
+};
+
+}}}
+
+namespace rclcpp { namespace subscription { namespace mock {
+
+class SubscriptionBase
+{
+public:
+  RCLCPP_SMART_PTR_DEFINITIONS(SubscriptionBase);
+
+  SubscriptionBase() : mock_topic_name(""), mock_queue_size(0) {}
+
+  std::string get_topic_name() const {return mock_topic_name;}
+  size_t get_queue_size() const {return mock_queue_size;}
+
+  std::string mock_topic_name;
+  size_t mock_queue_size;
+};
+
+}}}
+
+// Prevent rclcpp/publisher.hpp and rclcpp/subscription.hpp from being imported.
+#define RCLCPP_RCLCPP_PUBLISHER_HPP_
+#define RCLCPP_RCLCPP_SUBSCRIPTION_HPP_
+// Force ipm to use our mock publisher class.
+#define Publisher mock::Publisher
+#define SubscriptionBase mock::SubscriptionBase
+#include <rclcpp/intra_process_manager.hpp>
+#undef SubscriptionBase
+#undef Publisher
+
+#include <rcl_interfaces/msg/intra_process_message.hpp>
+
+/*
+This tests the "normal" usage of the class:
+ - Creates two publishers on different topics.
+ - Creates a subscription which matches one of them.
+ - Publishes on each publisher with different message content.
+ - Try's to take the message from the non-matching publish, should fail.
+ - Try's to take the message from the matching publish, should work.
+ - Asserts the message it got back was the one that went in (since there's only one subscription).
+ - Try's to take the message again, should fail.
+*/
+TEST(test_intra_process_manager, nominal) {
+  rclcpp::intra_process_manager::IntraProcessManager ipm;
+
+  auto p1 = std::make_shared<rclcpp::publisher::mock::Publisher>();
+  p1->mock_topic_name = "nominal1";
+  p1->mock_queue_size = 2;
+
+  auto p2 = std::make_shared<rclcpp::publisher::mock::Publisher>();
+  p2->mock_topic_name = "nominal2";
+  p2->mock_queue_size = 10;
+
+  auto s1 = std::make_shared<rclcpp::subscription::mock::SubscriptionBase>();
+  s1->mock_topic_name = "nominal1";
+  s1->mock_queue_size = 10;
+
+  auto p1_id = ipm.add_publisher<rcl_interfaces::msg::IntraProcessMessage>(p1);
+  auto p2_id = ipm.add_publisher<rcl_interfaces::msg::IntraProcessMessage>(p2);
+  auto s1_id = ipm.add_subscription(s1);
+
+  auto ipm_msg = std::make_shared<rcl_interfaces::msg::IntraProcessMessage>();
+  ipm_msg->message_sequence = 42;
+  ipm_msg->publisher_id = 42;
+  rcl_interfaces::msg::IntraProcessMessage::UniquePtr unique_msg;
+  unique_msg.reset(new rcl_interfaces::msg::IntraProcessMessage(*ipm_msg));
```

Comment:
Line length. And also because I was moving the first use of `unique_msg` around so it was easier to have two steps. Now that things have settled I can fix that: 58bc279